### ISSUE TITLE
Training freezes at the bad result and not improving

### ISSUE THREAD

**Issue by cxovrika** [Unity-Technologies/ml-agents#310]
Hello!

I am trying to train agent on a simple game: you are a capsule on a plane, you can move(WASD) and shoot upwards, dummy enemy is patrolling from left to the right of the board, and is shooting downwards every 2 sec.

GIF showing the game:  http://g.recordit.co/jJlwCkLni5.gif

while training, after every agentStep i give -10 reward for dying of falling off the platform, 20 reward for killing the enemy, -0.01 reward for trying to shoot while it is on cooldown and -0.005 otherwise.

the training goes really bad, as you can see from here: https://pastebin.com/t0zZxzJA
for some reason it goes from random moves(which give nearly 2.5 mean reward) to the moves that guarantee to give the -10 reward (STD of reward drops to almost zero as you can see)

when i gave the given bad model to the agent, for some reason that i don't understand it was moving only on the diagonal, from it's starting position:  http://g.recordit.co/LSXcEcSk47.gif

some additional info
hyperparameters: https://pastebin.com/iir3R4sH
info about brain: https://snag.gy/4lW0oN.jpg
state stores information about your, enemy's and enemy's projectile's position and velocity

any ideas why is this happening?

thanks in advance for any help!

**Comment by MarcoMeter** [comment]
Hi @cxovrika 

>  after every agentStep i give -10 reward for dying of falling off the platform, 20 reward for killing the enemy

Your chosen rewards a really high. I suspect that this causes a completely unstable training.  I'd suggest to try signaling rewards in the range -1 to +1.

> buffer_size = 409600

That's too much for sure. Most likely the buffer is likely to retain lots of bad training examples.

> batch_size = 2048

If you are using discrete actions, I made the experience for simple environments, like yours, to go with a much much smaller batch size like 16. Also your neural net structure is pretty high. There is not that much variety in your environment that it needs such a huge scale. Especially for only 10 inputs which describe the current state of the agent and the environment.

I just spotted that you are using a continuous action space. For your scenario, I'd go with discrete actions.  Shooting is not really a continuous action and WASD are typical key inputs. Also in continuous action space, actions are performed simultaneously.


One question, what information are you adding to the input space?

**Comment by cxovrika** [comment]
Thanks  for response @MarcoMeter 

i changed hyperparameters as you said, also i changed the rewards to the [-1; 1] range, normalized all inputs but still couldn't manage the proper training: https://pastebin.com/fqveUExf (also are these warnings important?)
i am using these new **hyperparameters**:  https://pastebin.com/5FMPNXbC
chcanged **buffer_size**, **batch_size** and **hidden_units**.

i wanna show you the codes for collectState and agentStep in case if i'm doing something wrong:
**collectState**: https://pastebin.com/XdBbaGb8
**agentStep**: https://pastebin.com/jqU3LVNY
all the parameters are in the [-10; 10] range. if there is no enemy fireball, then i'm adding some default 100-es to fill the list to the needed size (is this ok?)

Now i have only 3 actions, for horizontal movement, vertical movement and shooting. The reason i am doing it in a continuous action space type is because i want my next agent to be able to choose the direction of shooting, so i am gonna increase action size by 2, for direction's X and Z components, so the agent will be able to represent a direction vector from that two values(if this is a bad way of adding direction then please suggest a good one).

And also i have one question, if i'm using discrete action space of size 5, WASD for moving and space for shooting, and also for example i want the agent to be able to move left and shoot at the same time, in 1 frame, it won't be able to do this with that five actions right? i have to add  move+shoot action for the agent for every direction.

Thats all the information needed i think, thanks again zoro san!

**Comment by MarcoMeter** [comment]
Concerning the warning, it looks like that there might be an issue to save the model. If the trained model loads successfully, I'd maybe ignore it. I don't know too much about the underlying implementation to give a clear answer. Which version of tensorflow are you using?

If you are sticking to continues actions, I suggest to use the standard hyperparameters, which work for the 3DBall environment.

The final inputs fed to the neural net should be in the range [-1; 1]. Filling in the blanks with dummy data is alright, I'd rather go with -1 in your case.

Simultaneous actions in discrete space would be, like you said, having an action which comprises both actions. This ends up in a pretty high dimensional output vector.

**Comment by cxovrika** [comment]
Finally managed it to train properly, the problem was the same as in this issue #156 . Fixed it with some kind of a hack code changes as vincentpierre said. 

Thanks for all the help @MarcoMeter
